07-Concurrency_Synchronization/threads.c

```diff
@@ -0,0 +1,300 @@
+                ret = -ENOMEM;
+                goto out;
+        }
+
+        pr_info("thread: memory allocated\n");
+
+        return 0;
+
+out:
+        pr_err("thread : failed to allocate memory\n");
+        buffer_clean();
+        return ret;
+}
+
+static void remove_proc(void)
+{
+        if (proc_file) {
+                remove_proc_entry(PROC_FILENAME, proc_dir);
+                proc_file = NULL;
+        }
+
+        if (proc_dir) {
+                remove_proc_entry(PROC_DIR, NULL);
+                proc_dir = NULL;
+        }
+}
+
+static int create_proc(void)
+{
+        int ret;
+
+        proc_dir = proc_mkdir(PROC_DIR, NULL);
+        if (NULL == proc_dir) {
+                ret = -EFAULT;
+                goto out;
+        }
+
+        proc_file = proc_create(PROC_FILENAME, S_IFREG | S_IRUGO | S_IWUGO, proc_dir, &proc_fops);
+        if (NULL == proc_file) {
+                ret = -EFAULT;
+                goto out;
+        }
+
+        pr_info("thread: created procfs \n");
+
+        return 0;
+
+out:
+        pr_err("thread: failed to create procfs \n");
+        remove_proc();
+        return ret;
+}
+
+static int proc_write(struct file *filp, const char *user_buffer, size_t count, loff_t *offp)
+{
+        int ret;
+        size = count;
+
+        if (size > BUFFER_SIZE) {
+                size = BUFFER_SIZE;
+        }
+	mutex_lock(&lock);
+	ret = raw_copy_from_user(buff,user_buffer,size);
+	complete(&data);
+	mutex_unlock(&lock);
+        if (ret) {
+                ret = -EFAULT;
+                goto out;
+        }
+
+        return size;
+
+out:
+        return ret;
+}
+
+static int proc_read(struct file *filp, char *buffer, size_t len, loff_t *offp)
+{
+        int ret;
+        char msg[BUFFER_SIZE];
+
+        sprintf(msg, buff);
+
+        if (*offp >= strlen(msg)) {
+                *offp = 0;
+                ret = 0;
+                goto out;
+        }
+
+        if (len > strlen(msg) - *offp) {
+                len = strlen(msg) - *offp;
+        }
+	
+	mutex_lock(&lock);
+        ret = raw_copy_to_user((void*)buffer, msg - *offp, len);
+	mutex_unlock(&lock);
+
+        *offp += len;
+
+        return len;
+
+out:
+        return ret;
+}
+
+void __exit thread_exit(void)
+{
+
+	kthreads_stop();
+        remove_proc();
+
+        buffer_clean();
+
+        pr_info("thread : module exited\n");
+}
+
+int __init thread_init(void)
+{
+	
+	int ret;
+
+	ret = buffer_create();
+	if (ret) {
+		goto out;
+	}
+
+	proc_fops.read = proc_read;
+	proc_fops.write = proc_write;
+
+	create_proc();
+	ret = create_proc();
+	if (ret) {
+		goto out;
+        }
+
+	sema_init(&workersS, POOL_SIZE);
+	pr_info("thread : semaphore init\n");
+
+	mutex_init(&lock);
+
+	init_completion(&data);
+
+	masterT = kthread_run(master, NULL, "master_thread");
+
+	pr_info("thread : module loaded\n");
+
+        return 0;
+
+out:
+        return ret;
+}
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Rubinshteyn Mark,mark.rubinshteyn@meta.ua");
```

Comment:
```suggestion
MODULE_AUTHOR("Rubinshteyn Mark,mark.rubinshteyn@meta.ua");
MODULE_DESCRIPTION("Driver ???");


```

Please add  MODULE_DESCRIPTION